web/src/app/core/services/token.service.ts

```diff
@@ -0,0 +1,106 @@
+import { Injectable } from '@angular/core';
+import { throwError, Observable } from 'rxjs';
+import { first, map, switchMap } from 'rxjs/operators';
+import { v4 as uuid } from 'uuid';
+
+// Dashboard model and services
+import { IProjectTokenModel, ProjectModel } from '../../shared/models/index.model';
+import { ProjectService } from './project.service';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class TokenService {
+
+  constructor(
+    private projectService: ProjectService
+  ) { }
+
+  // This function returns the token details via id
+  public findOneById(projectUid: string, tokenUid: string): Observable<IProjectTokenModel> {
+    return this.projectService.findOneById(projectUid)
+      .pipe(
+        map((data: ProjectModel) => data && data.tokens && data.tokens.find((item: IProjectTokenModel) => item.uid === tokenUid))
```

Comment:
I think this can be shortened to as of typescript `v3.5` https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#non-null-assertion-operator

```typescript
map((data: ProjectModel) => data!.tokens!.find((item: IProjectTokenModel) => item.uid === tokenUid))
```

Reply:
The non-null assertion operator only for compilation step and does not check a variable on execute step

Reply:
Good point! 👍 . I think the sun got to me that day 🤦‍♂